Automattic/pocket-casts-android
# Prefer faster networks for downloads on watch

## Description
This PR uses the horologist network-awareness library to prefer faster networks when we download media. Without this, the watch defaults to using slower network connections (i.e., Bluetooth). More info [here](https://google.github.io/horologist/network-awareness/).

There was some trickiness integrating this because this wraps the `Call.Factory` which we create in the repository module. I initially tried using the library directly there, but the problem is that the network-awareness library (1) uses watch features (or at least claims it does [in its manifest file](https://github.com/google/horologist/blob/main/network-awareness/src/main/AndroidManifest.xml#L22)) and (2) has min SDK of 26. Since the phone app uses our repository module, both of these requirements created issues.

To work around this, I have created an annotation in the repository module to provide the relevant `Call.Factory`, but the 
 `@Provides` methods for those annotations are in the application modules (phone, automotive, and wear). I'm keeping the actual construction of the non-wear dependencies in the repository module—the application module is really only responsible for identifying which dependency to inject.

I had to do something similar with the `Request.Builder` because this library creates an extension method (`.requestType(...)` that needs to be set with an object from the library.

I'm only doing this for downloads currently, but I imagine I'll be doing something similar for some of our other okhttp requests in the future, so let me know any issues you see with this pattern.

This is a bit confusing from the DI perspective, but it completely avoids having any horologist dependency in our repository module, which feels like a nice win. With that said, let me know if you think this is too messy and not worth it.

## Testing Instructions
1. Install the app on a physical watch that is paired to a phone by Bluetooth and connected to a wifi network
2. Go to the episode details screen and download an episode
3. Check the logs for the watch, you should see a message along the lines of `HTTPS request media-download Wifi` reflecting that Wifi was used for the download
4. Alter the code to prefer Bluetooth connections for downloads [like this](https://gist.github.com/mchowning/6be13d811c8346ac849be44dd8806223).
5. Reinstall the app with the altered code
6. Download an episode again
7. Check the logs for the watch, you should see a message along the lines of `HTTPS request media-download BT` reflecting that Bluetooth was used for the download

## Checklist
